lib/db_utils.go

```diff
@@ -8573,50 +8605,66 @@ func DBGetPostAssociationByAttributesWithTxn(txn *badger.Txn, snap *Snapshot, qu
+		handle, snap, associationQuery, deletedUtxoAssociationIDs,
+	)
 	if err != nil {
-		return nil, errors.Wrapf(err, "DBGetUserAssociationsByAttributes: ")
+		return nil, nil, errors.Wrapf(err, "DBGetUserAssociationsByAttributes: ")
 	}
 
 	// Map from association IDs to association entries.
 	var associationEntries []*UserAssociationEntry
-
-	for _, associationID := range associationIDs {
+	for _, associationID := range associationIDs.ToOrderedSlice() {
 		// Retrieve association entry from db by ID.
 		associationEntry, err := DBGetUserAssociationByID(handle, snap, associationID)
 		if err != nil {
-			return nil, errors.Wrapf(err, "DBGetUserAssociationsByAttributes: problem retrieving association entry by ID: ")
+			return nil, nil, errors.Wrapf(err, "DBGetUserAssociationsByAttributes: problem retrieving association entry by ID: ")
 		}
 		associationEntries = append(associationEntries, associationEntry)
 	}
-	return associationEntries, nil
+	return associationEntries, prefixType, nil
 }
 
-func DBGetUserAssociationIdsByAttributes(handle *badger.DB, snap *Snapshot, associationQuery *UserAssociationQuery) ([]*BlockHash, error) {
+func DBGetUserAssociationIdsByAttributes(
+	handle *badger.DB,
+	snap *Snapshot,
+	associationQuery *UserAssociationQuery,
+	deletedUtxoAssociationIDs *Set[*BlockHash],
+) (*Set[*BlockHash], []byte, error) {
 	// Construct key based on input query params.
+	var prefixType []byte
 	var keyPrefix []byte
+	var err error
 
 	// TransactorPKID + TargetUserPKID
 	if associationQuery.TransactorPKID != nil {
 		if associationQuery.TargetUserPKID != nil {
 			// PrefixUserAssociationByUsers: TransactorPKID, TargetUserPKID, AssociationType, AssociationValue
+			prefixType = Prefixes.PrefixUserAssociationByUsers
 			keyPrefix = append(keyPrefix, Prefixes.PrefixUserAssociationByUsers...)
 			keyPrefix = append(keyPrefix, associationQuery.TransactorPKID.ToBytes()...)
 			keyPrefix = append(keyPrefix, associationQuery.TargetUserPKID.ToBytes()...)
 		} else {
 			// PrefixUserAssociationByTransactor: TransactorPKID, AssociationType, AssociationValue, TargetUserPKID
+			prefixType = Prefixes.PrefixUserAssociationByTransactor
 			keyPrefix = append(keyPrefix, Prefixes.PrefixUserAssociationByTransactor...)
 			keyPrefix = append(keyPrefix, associationQuery.TransactorPKID.ToBytes()...)
 		}
 	} else if associationQuery.TargetUserPKID != nil {
 		// PrefixUserAssociationByTargetUser: TargetUserPKID, AssociationType, AssociationValue, TransactorPKID
+		prefixType = Prefixes.PrefixUserAssociationByTargetUser
 		keyPrefix = append(keyPrefix, Prefixes.PrefixUserAssociationByTargetUser...)
 		keyPrefix = append(keyPrefix, associationQuery.TargetUserPKID.ToBytes()...)
 	} else {
 		// TransactorPKID == nil, TargetUserPKID == nil
-		return nil, errors.New("DBGetUserAssociationIdsByAttributes: invalid query params")
+		return nil, nil, errors.New("DBGetUserAssociationIdsByAttributes: invalid query params")
```

Comment:
I generally like to have my error cases up front. So `if TransactorPKID == nil && TargetUserPKID == nil { throw error }`. Reducing the number of `else` and `else if` statements is encouraged

Reply:
I generally agree, but appending these keys is kinda tricky since we don't break or return so have to use if/else if/elses to control logic flow. Here we have a `if A != nil else if B != nil else raise Error` which IMO is clearer than `if A == nil && B == nil raise Error; if A != nil else if B != nil` but LMK if you disagree.

Reply:
nah it's fine, leave it.